controllers/alertsconfig_controller.go

```diff
@@ -50,14 +71,212 @@ type AlertsConfigReconciler struct {
+		if err := r.Get(ctx, wfAlertNamespacedName, &wfAlert); err != nil {
+			log.Error(err, "unable to get the wavefront alert details for the requested name", "wfAlertName", config.AlertName)
+
+			// Update the status and retry it
+			return r.PatchIndividualAlertsConfigError(ctx, &alertsConfig, config.AlertName, err)
+		}
+		wfAlertBytes, err := json.Marshal(wfAlert.Spec)
+		if err != nil {
+			// update the status and retry it
+			return r.PatchIndividualAlertsConfigError(ctx, &alertsConfig, config.AlertName, err)
+		}
+
+		// execute Golang Template
+		wfAlertTemplate, err := template.ProcessTemplate(ctx, string(wfAlertBytes), config.Params)
+		if err != nil {
+			//update the status and retry it
+			return r.PatchIndividualAlertsConfigError(ctx, &alertsConfig, config.AlertName, err)
+		}
+		log.Info("Template process is successful", "here", wfAlertTemplate)
+
+		// Unmarshal back to wavefront alert
+		if err := json.Unmarshal([]byte(wfAlertTemplate), &wfAlert.Spec); err != nil {
+			// update the wfAlert status and retry it
+			return r.PatchIndividualAlertsConfigError(ctx, &alertsConfig, config.AlertName, err)
+		}
+		// Convert to Alert
+		var alert wf.Alert
+
+		if err := wavefront.ConvertAlertCRToWavefrontRequest(ctx, wfAlert.Spec, &alert); err != nil {
+			errMsg := "unable to convert the wavefront spec to Alert API request. will not be retried"
+			log.Error(err, errMsg)
+			return r.PatchIndividualAlertsConfigError(ctx, &alertsConfig, config.AlertName, err)
+		}
+
+		// Create/Update Alert
+		if alertHashMap[config.AlertName].ID == "" {
+			// Create use case
+			if err := r.WavefrontClient.CreateAlert(ctx, &alert); err != nil {
+				r.Recorder.Event(&alertsConfig, v1.EventTypeWarning, err.Error(), "unable to create the alert")
+				state := alertmanagerv1alpha1.Error
+				if strings.Contains(err.Error(), "Exceeded limit setting") {
+					// For ex: error is "Exceeded limit setting: 100 alerts allowed per customer"
+					state = alertmanagerv1alpha1.Error
+				}
+				log.Error(err, "unable to create the alert")
+
+				return r.CommonClient.UpdateStatus(ctx, &alertsConfig, state, errRequeueTime)
+			}
+			alertStatus := alertmanagerv1alpha1.AlertStatus{
+				ID:                 *alert.ID,
+				Name:               alert.Name,
+				LastChangeChecksum: reqChecksum,
+				Link:               fmt.Sprintf("https://%s/alerts/%s", internalconfig.Props.WavefrontAPIUrl(), *alert.ID),
+				State:              alertmanagerv1alpha1.Ready,
+				AssociatedAlert: alertmanagerv1alpha1.AssociatedAlert{
+					CR: config.AlertName,
+				},
+			}
+
+			alertStatusBytes, _ := json.Marshal(alertStatus)
+			patch := []byte(fmt.Sprintf("{\"status\":{\"state\": \"%s\", \"alertsCount\": 0, \"retryCount\": 0, \"alertStatus\":{\"%s\":%s}}}", alertmanagerv1alpha1.Ready, config.AlertName, string(alertStatusBytes)))
+			r.CommonClient.PatchStatus(ctx, &alertsConfig, client.RawPatch(types.MergePatchType, patch), alertmanagerv1alpha1.Ready)
```

Comment:
Yeah- Let me update it- Here is the reason why I moved out of StatusUpdate and started doing patch- Lets say you have 5 different alert configs in AlertsConfig CR and if you try to update that will try to update the entire status block and when the 2nd request goes the status object is already out of sync since first one changed the status or in the worst case 2nd one will overwrite the first one. Thats why patch is preferred since I'm touching only that specific alert part and not the entire status object.